ddtrace/internal/processor/stats.py

```diff
@@ -0,0 +1,258 @@
+                    "OkSummary": DDSketchProto.to_proto(stat_aggr.ok_distribution).SerializeToString(),
+                    "ErrorSummary": DDSketchProto.to_proto(stat_aggr.err_distribution).SerializeToString(),
+                }
+                if service:
+                    serialized_bucket["Service"] = service
+                if _type:
+                    serialized_bucket["Type"] = _type
+                bucket_aggr_stats.append(serialized_bucket)
+            serialized_buckets.append(
+                {
+                    "Start": bucket_time_ns,
+                    "Duration": self._bucket_size_ns,
+                    "Stats": bucket_aggr_stats,
+                }
+            )
+
+        # Clear out buckets that have been serialized
+        for key in serialized_bucket_keys:
+            del self._buckets[key]
+
+        return serialized_buckets
+
+    def _flush_stats(self, payload):
+        # type: (bytes) -> None
+        try:
+            conn = get_connection(self._agent_url, self._timeout)
+            conn.request("PUT", self._endpoint, payload, self._headers)
+            resp = get_connection_response(conn)
+        except Exception:
+            log.error("failed to submit span stats to the Datadog agent at %s", self._agent_endpoint, exc_info=True)
+            raise
+        else:
+            if resp.status == 404:
+                log.error(
+                    "Datadog agent does not support tracer stats computation, disabling, please upgrade your agent"
+                )
+                self._enabled = False
+                return
+            elif resp.status >= 400:
+                log.error(
+                    "failed to send stats payload, %s (%s) response from Datadog agent at %s",
+                    resp.status,
+                    resp.reason,
+                    self._agent_endpoint,
+                )
+            else:
+                log.info("sent %s to %s", _human_size(len(payload)), self._agent_endpoint)
+
+    def periodic(self):
+        # type: (...) -> None
+
+        with self._lock:
+            serialized_stats = self._serialize_buckets()
+
+        if not serialized_stats:
+            # No stats to report, short-circuit.
+            return
+        raw_payload = {"Stats": serialized_stats}  # type: Dict[str, Union[List[Dict], str]]
+        hostname = get_hostname()
+        if hostname:
+            raw_payload["Hostname"] = hostname
+        if config.env:
+            raw_payload["Env"] = config.env
+        if config.version:
+            raw_payload["Version"] = config.version
+
+        payload = packb(raw_payload)
+        try:
+            self._retry_request(self._flush_stats, payload)
+        except tenacity.RetryError:
+            log.error("retry limit exceeded submitting span stats to the Datadog agent at %s", self._agent_endpoint)
+
+    on_shutdown = periodic
+
+    def shutdown(self):
+        self.periodic()
+        self.stop()
```

Comment:
Do we need a try-except around ``self.stop``?  Do we want to handle the case where SpanStatsProcessorV06.shutdown() is called twice or it's called before the service is started

Reply:
We don't since the processor is started in the initializor. Unlike the AgentWriter when it can be initialized but not started.